File: backend/app/api/v1/projects.py
"""Project recommendations (Module 12)."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import CurrentUser, DBSession
from app.core.exceptions import ValidationError
from app.schemas.mvp import (
    ProjectCompleteRequest,
    ProjectCompletionsOut,
    ProjectRecommendationOut,
    ProjectSubmitRequest,
    ProjectSubmissionOut,
)
from app.services.analysis_service import AnalysisService
from app.services.project_evaluation_service import ProjectEvaluationService
from app.services.project_service import ProjectService

router = APIRouter()


def get_project_service(db: DBSession) -> ProjectService:
    return ProjectService(db)


def get_evaluation_service(db: DBSession) -> ProjectEvaluationService:
    return ProjectEvaluationService(db)


def _validate_title(title: str) -> str:
    cleaned = title.strip()
    if not cleaned:
        raise ValidationError("Le titre du projet est requis.")
    return cleaned


@router.get("/recommendations", response_model=ProjectRecommendationOut, summary="Recommended projects")
async def get_recommendations(
    current: CurrentUser,
    service: ProjectService = Depends(get_project_service),
) -> ProjectRecommendationOut:
    data = await service.recommend(current.id)
    return ProjectRecommendationOut.model_validate(data)


@router.get("/completions", response_model=ProjectCompletionsOut, summary="Completed portfolio projects")
async def list_completions(
    current: CurrentUser,
    service: ProjectService = Depends(get_project_service),
) -> ProjectCompletionsOut:
    completed = await service.list_completions(current.id)
    return ProjectCompletionsOut(completed=completed)


@router.post("/complete", response_model=ProjectCompletionsOut, summary="Mark project as completed")
async def complete_project(
    payload: ProjectCompleteRequest,
    current: CurrentUser,
    service: ProjectService = Depends(get_project_service),
) -> ProjectCompletionsOut:
    completed = await service.mark_complete(
        current.id, _validate_title(payload.title), payload.career_slug
    )
    return ProjectCompletionsOut(completed=completed)


@router.delete("/complete", response_model=ProjectCompletionsOut, summary="Unmark completed project")
async def uncomplete_project(
    payload: ProjectCompleteRequest,
    current: CurrentUser,
    service: ProjectService = Depends(get_project_service),
) -> ProjectCompletionsOut:
    completed = await service.mark_incomplete(current.id, _validate_title(payload.title))
    return ProjectCompletionsOut(completed=completed)


@router.post("/submit", response_model=ProjectSubmissionOut, summary="Submit project proof for evaluation")
async def submit_project_proof(
    payload: ProjectSubmitRequest,
    current: CurrentUser,
    evaluation_service: ProjectEvaluationService = Depends(get_evaluation_service),
    project_service: ProjectService = Depends(get_project_service),
) -> ProjectSubmissionOut:
    submission = await evaluation_service.submit(
        user_id=current.id,
        project_title=_validate_title(payload.project_title),
        career_slug=payload.career_slug,
        project_description=payload.project_description,
        skills_practiced=payload.skills_practiced,
        deliverables=payload.deliverables,
        github_url=payload.github_url or None,
        user_description=payload.user_description or None,
    )
    if submission.status == "approved":
        analysis_service = AnalysisService(project_service.db)
        await analysis_service.refresh_experience_level(current.id)
    return ProjectSubmissionOut.model_validate(submission)


@router.get("/submissions", response_model=list[ProjectSubmissionOut], summary="List my project submissions")
async def list_submissions(
    current: CurrentUser,
    evaluation_service: ProjectEvaluationService = Depends(get_evaluation_service),
) -> list[ProjectSubmissionOut]:
    submissions = await evaluation_service.get_submissions(current.id)
    return [ProjectSubmissionOut.model_validate(s) for s in submissions]


@router.get("/submissions/{project_title}", response_model=ProjectSubmissionOut | None, summary="Latest submission for a project")
async def get_project_submission(
    project_title: str,
    current: CurrentUser,
    evaluation_service: ProjectEvaluationService = Depends(get_evaluation_service),
) -> ProjectSubmissionOut | None:
    submission = await evaluation_service.get_latest_for_project(current.id, project_title)
    return ProjectSubmissionOut.model_validate(submission) if submission else None

File: frontend/src/services/projects.ts
import { api, apiSlow } from "@/services/api";
import type { ProjectRecommendation, ProjectSubmission } from "@/types";

export type { ProjectRecommendation, RecommendedProject, ProjectDataSource, ProjectSubmission } from "@/types";

export async function fetchProjectRecommendations(): Promise<ProjectRecommendation> {
  const { data } = await apiSlow.get<ProjectRecommendation>("/projects/recommendations");
  return data;
}

export async function fetchProjectCompletions(): Promise<string[]> {
  const { data } = await api.get<{ completed: string[] }>("/projects/completions");
  return data.completed;
}

export async function markProjectComplete(title: string, careerSlug?: string): Promise<string[]> {
  const { data } = await api.post<{ completed: string[] }>("/projects/complete", {
    title,
    career_slug: careerSlug ?? null,
  });
  return data.completed;
}

export async function unmarkProjectComplete(title: string): Promise<string[]> {
  const { data } = await api.delete<{ completed: string[] }>("/projects/complete", {
    data: { title },
  });
  return data.completed;
}

export interface SubmitProjectPayload {
  project_title: string;
  career_slug?: string | null;
  project_description?: string;
  skills_practiced?: string[];
  deliverables?: string[];
  github_url?: string | null;
  user_description?: string | null;
}

export async function submitProjectProof(payload: SubmitProjectPayload): Promise<ProjectSubmission> {
  const { data } = await api.post<ProjectSubmission>("/projects/submit", payload);
  return data;
}

export async function fetchProjectSubmissions(): Promise<ProjectSubmission[]> {
  const { data } = await api.get<ProjectSubmission[]>("/projects/submissions");
  return data;
}
